### PR TITLE
expect json response from the LLM

### PR DESCRIPTION
**Pull request by davidsbailey**
## Description

Adds option to handle a json response from the LLM instead of a TSV response. This improves accuracy and reliability for some LLMs including GPT 4 Turbo. 

The code changes in this PR enable running both the 'json' and 'reason' scenarios in the [LLM comparison](https://docs.google.com/spreadsheets/d/12bHzkrLx1kTbjUIf-hoSu3a3KIXFRMPqztDgB84-5CM/edit#gid=0) spreadsheet, for GPT 4 classic and turbo models.

### Code changes
* add response type param to params.json, which can either by `'tsv'` (default) or `'json'`.
* use response type to determine which filename suffix to reading example LLM responses from
* add `parse_json_response` method
* use response type decide whether to parse the LLM response as json

### S3 updates

alongside this PR, the following data has been uploaded to `s3://cdo-ai/teaching_assistant/experiments/`:
* `ai-rubrics-json`: copy of `ai-rubrics-pilot-baseline`, with:
  * system prompts updated to request JSON instead of TSV
  * L14 and L18 updated to provide JSON instead of TSV example responses
* `ai-rubrics-json-reason`: copy of `ai-rubrics-json`, with:
  * system prompt modified to request Reason before Grade, to improve chain-of-thought reasoning
  * L14 and L18 examples regenerated using GPT 4 classic and then hand-tuned for correctness (see below)
* `ai-rubrics-json-gpt-4-turbo`: copy of `ai-rubrics-json`, with `gpt-4-0125-preview` model
* `ai-rubrics-json-reason-gpt-4-turbo`: copy of `ai-rubrics-json-reason`, with `gpt-4-0125-preview` model

to help keep costs down and options open, I've included the `output/report-exact-match.html` files as well as the `cached_responses` directory. this will allow you, for example, to regenerate a pass-fail version of any report using rubric tester's `-c` option without incurring further LLM costs.

#### regenerating examples

as part of configuring `ai-rubrics-json-reason`, I used rubric tester to regenerate the examples (see steps added to the readme in this PR). Since I was regenerating existing examples rather than creating them from scratch, I chose the labels for the `actual_labels.csv` file in the temp dataset based on the `examples/*.tsv` files I was replacing, rather than having to determine those values from scratch. 

### rubric tester commands

The commands run to produce the results in [LLM comparison](https://docs.google.com/spreadsheets/d/12bHzkrLx1kTbjUIf-hoSu3a3KIXFRMPqztDgB84-5CM/edit#gid=0) are:
```
python ./lib/assessment/rubric_tester.py --experiment_name ai-rubrics-json
python ./lib/assessment/rubric_tester.py --experiment_name ai-rubrics-json-gpt-4-turbo
python ./lib/assessment/rubric_tester.py --experiment_name ai-rubrics-json-reason
python ./lib/assessment/rubric_tester.py --experiment_name ai-rubrics-json-reason-gpt-4-turbo
```
before repeating these commands, please note that each GPT 4 classic run costs about $12 and each GPT 4 Turbo run costs about $5, so the above 4 commands cost about $35 to run (!!). see previous note about how report outputs and cached responses have been uploaded to S3 to hopefully avoid costs of redundant test runs, or shrink the cost of your test runs with `-s` and `--lesson-names` rubric tester params.

## Testing story

* updated existing unit tests
* new unit test for reading json examples
* new unit test for parsing json response

### CODE REVIEW

File: tests/data/example.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+        "Grade": "Limited Evidence",
+        "Reason": "2 sprites are placed on the screen using the coordinate system, but no shapes or lines of text are created. At least 1 element moves during the program."
+    },
+    {
+        "Key Concept": "Optional \u201cStretch\u201d Feature - Variables",
+        "Observations": "(1) The variables created are: backgroundSprite, snowman, santa. The values stored in these variables are sprites. (2) No variables are created that are not used to store a sprite or a sprite property.",
+        "Grade": "No Evidence",
+        "Reason": "No non-sprite variables are created."
+    }
+]
```
Comment (davidsbailey) [inline tests/data/example.json:20]: Work in progress: need to add a unit tests which tests against this file.

File: tests/unit/assessment/test_rubric_tester.py [hunk 2]
```diff
@@ -178,14 +179,42 @@ def file_open_mock(name, *a, **kw):
...
 
         assert len(result) == len(examples_set)
 
         for i, example in enumerate(result):
             assert example[0] == examples_set[i][0]
             assert example[1] == examples_set[i][1]
 
+    def test_should_open_example_js_and_json_files(self, mocker, code_generator, rubric, examples):
+        examples_set = examples(rubric)
+        print(examples_set)
```
Comment (wilkie) [inline tests/unit/assessment/test_rubric_tester.py:192]: I think this is a stray print.

**Review by wilkie — APPROVED** [review]
Wow this is great stuff! I like how you've added the response type as another option... JSON will be so much better to work with overall.

I think I found one stray print, but other than that it looks great.

File: tests/unit/assessment/test_rubric_tester.py (hunk 2 above)
Comment (davidsbailey) [reply tests/unit/assessment/test_rubric_tester.py:192]: thank you! removed here and above

**Comment by davidsbailey** [comment]
just wanted to add, thank you @wilkie for all the great test coverage in here! as a python newbie, not having to figure out how to test my code changes from scratch has been huge help 😁

**Pull request merged**